mythril/laser/ethereum/state.py

```diff
@@ -1,16 +1,80 @@
-from z3 import BitVec, BitVecVal, Solver, ExprRef, sat
+from z3 import BitVec, BitVecVal, BitVecRef, BitVecNumRef, BitVecSort, Solver, ExprRef, Concat, sat, simplify, Array, ForAll, Solver, UGT, Implies
+from z3.z3types import Z3Exception
 from mythril.disassembler.disassembly import Disassembly
 from copy import copy, deepcopy
 from enum import Enum
 from random import randint
+from mythril.laser.ethereum.util import get_concrete_int
 
 from mythril.laser.ethereum.evm_exceptions import StackOverflowException, StackUnderflowException
 
-
 class CalldataType(Enum):
     CONCRETE = 1
     SYMBOLIC = 2
 
+class Calldata:
+
+    def __init__(self, tx_id, starting_calldata: bytes=None):
+        self.tx_id = tx_id
+        if starting_calldata:
+            self._calldata = []
+            self.calldatasize = BitVecVal(len(starting_calldata), 256)
+            self.concrete = True
+        else:
+            self._calldata = Array('{}_calldata'.format(self.tx_id), BitVecSort(256), BitVecSort(8))
+            self.calldatasize = BitVec('{}_calldatasize'.format(self.tx_id), 256)
+            self.concrete = False
+
+        self.starting_calldata = starting_calldata if starting_calldata else []
+
+    def set_global_state(self, state):
+        self.state = state
+        constraints = []
+
+        if self.concrete:
+            for cd_byte in self.starting_calldata:
+                if type(cd_byte) == int:
+                    self._calldata.append(BitVecVal(cd_byte, 8))
+                else:
+                    self._calldata.append(cd_byte)
+            constraints.append(self.calldatasize == len(self.starting_calldata))
+        else:
+            x = BitVec('x', 256)
+            constraints.append(ForAll(x, Implies(self[x] != 0, UGT(self.calldatasize, x))))
+
+        self.state.mstate.constraints.extend(constraints)
+
+    def concretized(self, model):
+        result = []
+        for i in range(get_concrete_int(model.eval(self.calldatasize, model_completion=True))):
+            result.append(get_concrete_int(model.eval(self[i], model_completion=True)))
+
+        return result
+
+    def get_word_at(self, index: int):
+        return self[index:index+32]
+
+    def __getitem__(self, item):
+        if isinstance(item, slice):
+            MAX_SLICE = 1024
+            try:
+                current_index = item.start if type(item.start) in [BitVecRef, BitVecNumRef] else BitVecVal(item.start, 256)
+                dataparts = []
+                i = 0
+                while simplify(current_index != item.stop):
+                    dataparts.append(self[current_index])
```

Comment:
Sure, I'll add something like that if the calldatasize is concrete. (I'll grab the calldatasize directly instead of finding the length)

I can't remember why I added the MAX_SLICE check... I think that was from when I was worried about things taking a long time, but I don't think that it is needed either, and it's obviously not being reached, so I'll remove that too.

Reply:
this will still result in the same problem right? as the the index might not be an int.

Reply:
this will still result in the same problem right? as the index may not be an int(say some BitVecNum or something of that sort which is in a way an int) ?

Reply:
this might still result in the same problem right? as the index can be some sort of BitVecNum or something which is `int` in a way?

Reply:
You're right, item.stop could be a BitVecVal. Good catch.

Reply:
Alright, it should be good now.